[CGP] Sink and duplicate call for tail position when duplicating return

antoniofrighetto opened the pull request:
Prior to duplicating returns in `dupRetToEnableTailCallOpts`, on the chance no tail call opportunities have been found yet, for triangle-like if/then it may be profitable to sink the call in one successor, and duplicate the call in the return block, so that this latter call may be eligible for tail call.

Fixes: https://github.com/llvm/llvm-project/issues/163811.

dtcxzyw commented:
Can you share some performance measurements about this optimization (e.g., llvm-test-suite/SPEC)? This pattern is common so the impact should be significant.

weiguozhi commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -3084,24 +3084,87 @@ bool CodeGenPrepare::dupRetToEnableTailCallOpts(BasicBlock *BB,
+    auto *OrigCITerm = cast<BranchInst>(OrigCI->getParent()->getTerminator());
+    auto *Cond = OrigCITerm->getCondition();
+    if (!isGuaranteedNotToBeUndefOrPoison(Cond)) {
+      auto *FI = new FreezeInst(Cond, Cond->getName() + ".fr",
```
Why do we need this?

rnk reviewed (changes requested):
> Can you share some performance measurements about this optimization (e.g., llvm-test-suite/SPEC)? This pattern is common so the impact should be significant.

I share these concerns, and I wonder if this would result in unprofitable transformations for calls with many arguments:
https://godbolt.org/z/4ej9xvebr

That example includes arguments passed in memory, but note how we get duplicate constant code generation. That's more of an underlying LLVM deficiency. We could do some post-regalloc pass to find branches with available physical registers with constant materialization on both sides to hoist... but we don't. We probably do better when the arguments are all virtual registers that need to get forwarded.

I would revisit this after there is some reasonable profitability heuristic in place.

antoniofrighetto committed:
[CGP] Introduce test for PR172882 (NFC)

antoniofrighetto committed:
[CGP] Sink and duplicate call for tail position when duplicating return

Prior to duplicating returns in `dupRetToEnableTailCallOpts`, on the
chance no tail call opportunities have been found yet, for triangle-
like if/then it may be profitable to sink the call in one successor,
and duplicate the call in the return block, so that this latter call
may be eligible for tail call.

Fixes: https://github.com/llvm/llvm-project/issues/163811.

antoniofrighetto committed:
!fixup list transformed code
